### ISSUE TITLE
Is the Certbot snap relying on bash from outside of the snap?

### ISSUE THREAD

**Issue by bmw** [certbot/certbot#8251]
In a similar vein to the problems discussed at https://github.com/certbot/certbot/issues/8245#issuecomment-684109842, is `certbot.wrapper` being run by `bash` on the host system? I haven't dug into this, but I suspect it is with the shebang `#!/bin/bash`.

If we are relying on bash on the host, I think we should fix it. While rare, not all systems have `bash`.

Fixing this may be somewhat tricky though, especially if we need `LD_LIBRARY_PATH` to be properly set before invoking `bash` inside of the snap. See [this code](https://github.com/certbot/certbot/blob/d62d853ea4c03676f1a10248d01b91aa61eaec3e/certbot.wrapper#L4-L31).

We could maybe alternatively rely on `/bin/sh` on the host, but I think containing this in the consistent snap environment should be preferred.

**Comment by bmw** [comment]
Another option for us here might be to use `/usr/bin/env <whatever executable we want from our PATH>`.

@alexzorin may already know the answer to this based on their work in https://github.com/certbot/certbot/pull/8249, but I wonder if our hooks are a problem too. Are they run with our normal "classic" confinement or is their environment more restrictive?

**Comment by adferrand** [comment]
Hooks execution are done in an environment cleaned of snap customization since #8108

**Comment by bmw** [comment]
Sorry I should have been more clear. I think we're talking about two different hooks.

You're talking about Certbot's hook flags while I'm talking about the snap hooks under [`snap/hooks/`](https://github.com/certbot/certbot/tree/3ce87d1fcb1a11b554c9e22727d602cb86f0b8bf/snap/hooks).

**Comment by bmw** [comment]
I happened to think that while I think we can use `#!/usr/bin/env [bash|python(3)]` for `certbot.wrapper` if we wanted, I'm not sure if we can use it for the snap hooks because I'm not sure what the environment those hooks are run in looks like and if we're able to modify it like we do for `certbot.wrapper` [here](https://github.com/certbot/certbot/blob/3ce87d1fcb1a11b554c9e22727d602cb86f0b8bf/snap/snapcraft.yaml#L23-L27).

**Comment by bmw** [comment]
Another place I think we want to fix things here are the plugin hooks that are generated by this script: https://github.com/certbot/certbot/blob/master/tools/snap/generate_dnsplugins_postrefreshhook.sh

**Comment by alexzorin** [comment]
This is the only leftover reference to bash I can find (besides some augeas lens test fixtures):

    $ ag "/bin/bash" /snap/certbot*/current
    /snap/certbot/current/usr/bin/sotruss
    1:#! /bin/bash

Not too sure what it's doing in there snap, but hopefully snap does not use it.

**Comment by bmw** [comment]
Nice! I'm closing this issue as fixed by https://github.com/certbot/certbot/pull/8249.

I still personally would slightly prefer if `certbot.wrapper` was written in Python rather than `/bin/sh` since I think it is more readable and would be easier to maintain than ensuring that the script works with everyone's `/bin/sh`. With certbot-auto, we found that just being POSIX complaint wasn't enough due to shells like Busybox. See https://github.com/certbot/certbot/pull/1901.

I'm not particularly worried about this right now though. Instead of maintaining an issue for this, I opened https://github.com/certbot/certbot/pull/8270 to add a comment about this in `certbot.wrapper` for people to see if we work on the script in the future.
